README.md
## r_upgrade_ui
[![pub package](https://img.shields.io/pub/v/r_upgrade_ui.svg)](https://pub.dartlang.org/packages/r_upgrade_ui)

[r_upgrade](https://pub.dev/packages/r_upgrade) UI Packages,simple to use.

## How to use?

1. add plugin to your `pubspec.yaml`

```yaml
dependencies:
  r_upgrade_ui: last_version
```

2. Create RUpgradeInfo class.

```dart

RUpgradeInfo rUpgradeInfo = RUpgradeInfo()
  ..androidInfo = (RUpgradeAndroidInfo()
    ..webUrl = 'http://www.baidu.com'
    ..downloadUrl =
        'https://mydata-1252536312.cos.ap-guangzhou.myqcloud.com/r_upgrade.apk'
    ..downloadFileName = 'r_upgrade.apk'
    ..downloadFinishAutoInstall = true
    ..downloadUseCache = false
    ..upgradeFlavor = RUpgradeFlavor.normal
    ..notificationVisibility = NotificationVisibility.VISIBILITY_VISIBLE
    ..notificationStyle = NotificationStyle.planTime)
  ..iosInfo = (RUpgradeIOSInfo()
    ..webUrl = 'http://www.google.com'
    ..isChina = false
    ..appId = '414478124')
  ..isForce = true
  ..newVersion = 'v1.0.1'
  ..newVersionCode = 1
  ..content = '1.fixed bug,\n2.fixed bug again'
  ..title = 'New Version';
```

3. Use Normal Download Dialog.

```dart
    rUpgradeInfo.showNormalDialog(context);
```

4. android upgrade method.

```dart
  // upgrade from store
rUpgradeInfo.androidInfo.upgradeFromStore();

// upgrade from download
rUpgradeInfo.androidInfo.upgradeFromDownload();
```

5. ios upgrade method.

```dart
  // upgrade from store
rUpgradeInfo.iosInfo.upgradeFromStore();

```

6. share method.

```dart
  // upgrade from store
rUpgradeInfo.upgradeFromStore();

// upgrade from url
rUpgradeInfo.upgradeFromUrl();
```

## LICENSE

    Copyright 2021 The r_upgrade_ui Project Authors

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


CHANGELOG.md
## 0.1.3

* upgrade flutter version 3.0
## 0.1.2

* upgrde r_upgrade packages version, fix [normalDialog] max_length == -1
## 0.1.1

* upgrade version.
## 0.1.0

* first commit.



lib/r_upgrade_ui.dart
library r_upgrade_ui;

export 'r_upgrade_info.dart';
export 'package:r_upgrade/r_upgrade.dart';
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:r_upgrade/r_upgrade.dart';

import 'r_upgrade_info.dart';

const _kAnimationDuration = 300;

class RUpgradeNormalDialog extends StatefulWidget {
  final RUpgradeInfo info;

  const RUpgradeNormalDialog({Key? key, required this.info}) : super(key: key);

  @override
  _RUpgradeNormalDialogState createState() => _RUpgradeNormalDialogState();
}

class _RUpgradeNormalDialogState extends State<RUpgradeNormalDialog>
    with TickerProviderStateMixin {
  late AnimationController _controller;

  bool enterNewStatus = false;
  bool enterNewStatus2 = false;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
        vsync: this,
        duration: const Duration(milliseconds: _kAnimationDuration));
    _controller.forward();
  }

  @override
  Widget build(BuildContext context) {
    return Material(
      type: MaterialType.transparency,
      child: Center(
        child: AnimatedContainer(
          duration: const Duration(milliseconds: _kAnimationDuration),
          width: 300,
          clipBehavior: Clip.antiAlias,
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(8),
            color: Theme.of(context).dialogBackgroundColor,
          ),
          child: enterNewStatus
              ? _buildAndroidUpgradeProgress(context)
              : _buildUpgradeContent(context),
        ),
      ),
    );
  }

  Widget _buildAndroidUpgradeProgress(BuildContext context) =>
      AnimatedContainer(
        duration: const Duration(milliseconds: _kAnimationDuration),
        height: enterNewStatus2 ? 120 : 250,
        child: Stack(
          children: [
            Positioned(
              top: 12,
              left: 16,
              child: FadeTransition(
                opacity: _controller,
                child: StreamBuilder(
                  stream: widget.info.androidInfo.downloadStream(),
                  builder: (BuildContext context,
                      AsyncSnapshot<DownloadInfo> snapshot) {
                    String status = 'Unknown';
                    bool showContinue = false;

                    if (snapshot.hasData) {
                      if (snapshot.data!.status ==
                          DownloadStatus.STATUS_SUCCESSFUL) {
                        status = 'Download Finish';
                        Navigator.of(context).pop();
                      } else if (snapshot.data!.status ==
                          DownloadStatus.STATUS_FAILED) {
                        status = 'Download Failed';
                        showContinue = true;
                      } else if (snapshot.data!.status ==
                          DownloadStatus.STATUS_CANCEL) {
                        status = 'Download Canceled';
                        showContinue = true;
                      } else if (snapshot.data!.status ==
                          DownloadStatus.STATUS_PENDING) {
                        status = 'Download Pending';
                      }
                      if (snapshot.data!.status ==
                          DownloadStatus.STATUS_RUNNING) {
                        status = 'Downloading';
                      }
                      if (snapshot.data!.status ==
                          DownloadStatus.STATUS_PAUSED) {
                        status = 'Downloading Paused';
                        showContinue = true;
                      }
                    }
                    return RichText(
                      text: TextSpan(
                        text: status,
                        children: [
                          if (showContinue)
                            TextSpan(
                                text: ' Continue',
                                style: TextStyle(
                                  color: Theme.of(context).primaryColor,
                                ),
                                recognizer: TapGestureRecognizer()
                                  ..onTap = () {
                                    widget.info.androidInfo
                                        .upgradeFromDownload();
                                  })
                        ],
                        style: Theme.of(context).textTheme.bodyText1,
                      ),
                    );
                  },
                ),
              ),
            ),
            Positioned(
              bottom: 16,
              left: 16,
              right: 16,
              child: FadeTransition(
                opacity: _controller,
                child: StreamBuilder(
                  stream: widget.info.androidInfo.downloadStream(),
                  builder: (BuildContext context,
                      AsyncSnapshot<DownloadInfo> snapshot) {
                    return Row(
                      mainAxisAlignment: MainAxisAlignment.spaceBetween,
                      children: [
                        Text(
                          snapshot.hasData
                              ? snapshot.data!.getSpeedString()
                              : '--mb/s',
                          style: Theme.of(context).textTheme.bodyText1,
                        ),
                        Visibility(
                          child: Text(
                            snapshot.hasData
                                ? '${snapshot.data!.percent!.toStringAsFixed(1)} %'
                                : '0.0%',
                            style: Theme.of(context).textTheme.bodyText1,
                          ),
                          visible:
                              snapshot.hasData && snapshot.data!.percent! > 0,
                        ),
                      ],
                    );
                  },
                ),
              ),
            ),
            AnimatedPositioned(
              onEnd: () {
                _controller.forward();
              },
              duration: const Duration(milliseconds: _kAnimationDuration),
              top: enterNewStatus2 ? 45 : 0,
              left: enterNewStatus2 ? 16 : 0,
              right: enterNewStatus2 ? 16 : 0,
              child: StreamBuilder<DownloadInfo>(
                  stream: widget.info.androidInfo.downloadStream(),
                  builder: (context, AsyncSnapshot<DownloadInfo> snapshot) {
                    if (snapshot.hasData && snapshot.data!.maxLength == -1) {
                      return AnimatedContainer(
                        height: enterNewStatus2 ? 30 : 40,
                        duration:
                            const Duration(milliseconds: _kAnimationDuration),
                        clipBehavior: Clip.antiAlias,
                        decoration: BoxDecoration(
                            color: enterNewStatus2
                                ? Colors.white
                                : Theme.of(context).primaryColor,
                            borderRadius:
                                BorderRadius.circular(enterNewStatus2 ? 8 : 0)),
                        child: LinearProgressIndicator(
                          minHeight: enterNewStatus2 ? 30 : 40,
                        ),
                      );
                    }
                    return AnimationShaderMask(
                      stops: const [1, 0],
                      colors: [
                        Theme.of(context).primaryColor,
                        Theme.of(context).primaryColor.withOpacity(0.2),
                      ],
                      progress: snapshot.hasData
                          ? (snapshot.data!.status ==
                                  DownloadStatus.STATUS_SUCCESSFUL
                              ? 1
                              : snapshot.data!.percent! / 100)
                          : 0,
                      duration:
                          const Duration(milliseconds: _kAnimationDuration),
                      child: AnimatedContainer(
                        height: enterNewStatus2 ? 30 : 40,
                        duration:
                            const Duration(milliseconds: _kAnimationDuration),
                        padding: const EdgeInsets.symmetric(
                          horizontal: 12,
                        ),
                        alignment: Alignment.center,
                        decoration: BoxDecoration(
                            color: enterNewStatus2
                                ? Colors.white
                                : Theme.of(context).primaryColor,
                            borderRadius:
                                BorderRadius.circular(enterNewStatus2 ? 8 : 0)),
                      ),
                    );
                  }),
            ),
          ],
        ),
      );

  Widget _buildUpgradeContent(BuildContext context) {
    return Column(
      mainAxisSize: MainAxisSize.min,
      children: [
        Container(
          height: 40,
          color: Theme.of(context).primaryColor,
          padding: const EdgeInsets.symmetric(
            horizontal: 12,
          ),
          alignment: Alignment.centerLeft,
          child: FadeTransition(
            opacity: _controller,
            child: Text(
              '${widget.info.title!}(${widget.info.newVersion}-${widget.info.newVersionCode})',
              style: Theme.of(context).primaryTextTheme.headline6!,
              maxLines: 1,
              overflow: TextOverflow.ellipsis,
            ),
          ),
        ),
        Container(
          height: 160,
          padding: const EdgeInsets.symmetric(
            horizontal: 12,
            vertical: 8,
          ),
          alignment: Alignment.topLeft,
          child: FadeTransition(
            opacity: _controller,
            child: SlideTransition(
              position: Tween<Offset>(
                      begin: const Offset(0, -0.5), end: const Offset(0, 0))
                  .animate(_controller),
              child: SingleChildScrollView(
                child: Text(
                  widget.info.content!,
                  style: Theme.of(context).textTheme.headline6,
                ),
              ),
            ),
          ),
        ),
        FadeTransition(
          opacity: _controller,
          child: Container(
            height: 0.5,
            color: Theme.of(context).dividerColor,
          ),
        ),
        FadeTransition(
          opacity: _controller,
          child: SlideTransition(
            position: Tween<Offset>(
                    begin: const Offset(0, 0.5), end: const Offset(0, 0))
                .animate(_controller),
            child: Row(
              children: [
                if (!(widget.info.isForce == true))
                  Expanded(
                      child: TextButton(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    child: const Text('Cancel'),
                  )),
                if (!(widget.info.isForce == true))
                  FadeTransition(
                    opacity: _controller,
                    child: Container(
                      width: 1,
                      height: 24,
                      color: Theme.of(context).dividerColor,
                    ),
                  ),
                Expanded(
                    child: TextButton(
                  onPressed: _onUpgrade,
                  child: const Text('Upgrade'),
                )),
              ],
            ),
          ),
        ),
      ],
    );
  }

  void _onAndroidDownload() async {
    bool isFinish = await widget.info.androidInfo.checkIsFinish();
    if (!isFinish || widget.info.androidInfo.downloadUseCache == false) {
      final future = _controller.reverse();
      future.whenComplete(() async {
        setState(() {
          enterNewStatus = true;
        });
        widget.info.androidInfo.upgradeFromDownload();
        WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
          setState(() {
            enterNewStatus2 = true;
          });
        });
      });
    } else {
      widget.info.androidInfo.upgradeFromDownload();
    }
  }

  void _onUpgrade() async {
    if (Platform.isAndroid) {
      if (widget.info.androidInfo.downloadUrl != null) {
        _onAndroidDownload();
      } else if (widget.info.androidInfo.webUrl != null) {
        await widget.info.upgradeFromUrl();
      } else {
        widget.info.upgradeFromStore();
      }
    } else if (Platform.isIOS) {
      if (widget.info.iosInfo.appId != null) {
        await widget.info.upgradeFromStore();
      } else {
        await widget.info.upgradeFromUrl();
      }
    }
  }
}

class AnimationShaderMask extends ImplicitlyAnimatedWidget {
  /// Creates a widget that animates its opacity implicitly.
  ///
  /// The [progress] argument must not be null and must be between 0.0 and 1.0,
  /// inclusive. The [curve] and [duration] arguments must not be null.
  const AnimationShaderMask({
    Key? key,
    this.child,
    required this.progress,
    required this.colors,
    this.stops,
    Curve curve = Curves.linear,
    required Duration duration,
    VoidCallback? onEnd,
    this.alwaysIncludeSemantics = false,
  })  : assert(progress >= 0.0 && progress <= 1.0),
        super(key: key, curve: curve, duration: duration, onEnd: onEnd);

  final Widget? child;
  final double progress;
  final bool alwaysIncludeSemantics;
  final List<Color> colors;
  final List<double>? stops;

  @override
  ImplicitlyAnimatedWidgetState<AnimationShaderMask> createState() =>
      _AnimationShaderMaskState();

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(DoubleProperty('progress', progress));
  }
}

class _AnimationShaderMaskState
    extends ImplicitlyAnimatedWidgetState<AnimationShaderMask> {
  Tween<double>? _progress;
  late Animation<double> _opacityAnimation;

  @override
  void forEachTween(TweenVisitor<dynamic> visitor) {
    _progress = visitor(_progress, widget.progress,
            (dynamic value) => Tween<double>(begin: value as double))
        as Tween<double>?;
  }

  @override
  void didUpdateTweens() {
    _opacityAnimation = animation.drive(_progress!);
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _opacityAnimation,
      builder: (BuildContext context, Widget? child) {
        return ShaderMask(
          shaderCallback: (rect) {
            return LinearGradient(colors: widget.colors, stops: widget.stops)
                .createShader(Rect.fromLTWH(
                    0, 0, rect.width * _opacityAnimation.value, rect.height));
          },
          child: child,
        );
      },
      child: widget.child,
    );
  }
}


lib/r_upgrade_info.dart
import 'dart:io';

import 'package:flutter/material.dart';

import 'r_upgrade_ui.dart';

class RUpgradeAndroidInfo {
  RUpgradeFlavor upgradeFlavor = RUpgradeFlavor.normal;

  Map<String, String>? downloadHeader;

  // you want to download url.
  String? downloadUrl;

  // you want to download file name
  String? downloadFileName;

  // Is finish download? ** read only
  bool? _downloadIsFinish;

  // auto install when download finish.
  bool? downloadFinishAutoInstall;

  // Is finish download? ** read only
  bool? get downloadIsFinish => _downloadIsFinish;

  // android store id.
  String? storeId;

  // web download url
  String? webUrl;

  // download id ** read only
  int? _downloadId;

  // download id ** read only
  int? get downloadId => _downloadId;

  // use last upgrade cache from this version.
  bool downloadUseCache = true;

  // notification visibility
  NotificationVisibility? notificationVisibility;

  // notification style
  NotificationStyle? notificationStyle;

  // upgrade from store
  Future<bool?> upgradeFromStore() async {
    List<AndroidStore>? storeList = await RUpgrade.androidStores;
    if (storeList != null && storeList.isNotEmpty) {
      AndroidStore store = storeList.first;
      for (final item in storeList) {
        final version = await RUpgrade.getVersionFromAndroidStore(item);
        if (version != null) {
          store = item;
          break;
        }
      }
      return await RUpgrade.upgradeFromAndroidStore(store);
    }
    return false;
  }

  // check is finish
  Future<bool> checkIsFinish() async {
    _downloadIsFinish = false;
    _downloadId = await RUpgrade.getLastUpgradedId();
    if (_downloadId != null) {
      final status = await RUpgrade.getDownloadStatus(_downloadId!);
      if (status == DownloadStatus.STATUS_SUCCESSFUL) {
        _downloadIsFinish = true;
        return true;
      }
    }
    return false;
  }

  // upgrade form download
  Future<bool> upgradeFromDownload() async {
    bool shouldRestartDownload = false;
    _downloadIsFinish = false;
    if (downloadUseCache != true) {
      shouldRestartDownload = true;
    } else {
      _downloadId = await RUpgrade.getLastUpgradedId();
      if (_downloadId != null) {
        final status = await RUpgrade.getDownloadStatus(_downloadId!);
        if (status == DownloadStatus.STATUS_SUCCESSFUL) {
          _downloadIsFinish = true;
          await RUpgrade.upgradeWithId(_downloadId!);
        } else if (status == DownloadStatus.STATUS_FAILED) {
          shouldRestartDownload = true;
        } else {
          await RUpgrade.upgradeWithId(_downloadId!);
        }
      } else {
        shouldRestartDownload = true;
      }
    }

    if (shouldRestartDownload) {
      assert(downloadUrl != null, 'Please set android downloadUrl');
      _downloadId = await RUpgrade.upgrade(
        downloadUrl!,
        header: downloadHeader,
        fileName: downloadFileName,
        notificationVisibility:
            notificationVisibility ?? NotificationVisibility.VISIBILITY_VISIBLE,
        notificationStyle: notificationStyle ?? NotificationStyle.planTime,
        isAutoRequestInstall: downloadFinishAutoInstall ?? true,
        upgradeFlavor: upgradeFlavor,
      );
    }
    return false;
  }

  //download stream
  Stream<DownloadInfo> downloadStream() => RUpgrade.stream;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is RUpgradeAndroidInfo &&
          runtimeType == other.runtimeType &&
          upgradeFlavor == other.upgradeFlavor &&
          downloadHeader == other.downloadHeader &&
          downloadUrl == other.downloadUrl &&
          downloadFileName == other.downloadFileName &&
          _downloadIsFinish == other._downloadIsFinish &&
          downloadFinishAutoInstall == other.downloadFinishAutoInstall &&
          storeId == other.storeId &&
          webUrl == other.webUrl &&
          _downloadId == other._downloadId &&
          notificationVisibility == other.notificationVisibility &&
          notificationStyle == other.notificationStyle;

  @override
  int get hashCode =>
      upgradeFlavor.hashCode ^
      downloadHeader.hashCode ^
      downloadUrl.hashCode ^
      downloadFileName.hashCode ^
      _downloadIsFinish.hashCode ^
      downloadFinishAutoInstall.hashCode ^
      storeId.hashCode ^
      webUrl.hashCode ^
      _downloadId.hashCode ^
      notificationVisibility.hashCode ^
      notificationStyle.hashCode;

  @override
  String toString() {
    return 'RUpgradeAndroidInfo{upgradeFlavor: $upgradeFlavor, downloadHeader: $downloadHeader, downloadUrl: $downloadUrl, downloadFileName: $downloadFileName, _downloadIsFinish: $_downloadIsFinish, downloadFinishAutoInstall: $downloadFinishAutoInstall, storeId: $storeId, webUrl: $webUrl, _downloadId: $_downloadId, notificationVisibility: $notificationVisibility, notificationStyle: $notificationStyle}';
  }
}

class RUpgradeIOSInfo {
  // app store id
  String? appId;

  // web download url
  String? webUrl;

  // is use china appstore?
  bool isChina = true;

  Future<bool?> upgradeFromStore() async {
    return await RUpgrade.upgradeFromAppStore(appId!, isChina);
  }
}

class RUpgradeInfo {
  late RUpgradeAndroidInfo androidInfo;

  late RUpgradeIOSInfo iosInfo;

  // force upgrade?
  bool? isForce;

  // upgrade from background?
  bool? isBackground;

  // upgrade title
  String? title;

  // upgrade content
  String? content;

  // new version name
  String? newVersion;

  // new version code
  int? newVersionCode;

  // web url is null.
  bool get webUrlIsNull => androidInfo.webUrl == null && iosInfo.webUrl == null;

  String getWebUrl() {
    String? webUrl;
    if (Platform.isAndroid) {
      webUrl = androidInfo.webUrl;
    } else if (Platform.isIOS) {
      webUrl = iosInfo.webUrl;
    }
    assert(webUrl != null, 'Web url can not null!');
    return webUrl!;
  }

  Future<bool?> upgradeFromUrl() async {
    if (!webUrlIsNull) {
      return await RUpgrade.upgradeFromUrl(getWebUrl());
    }
    return false;
  }

  Future<bool?> upgradeFromStore() async {
    if (Platform.isAndroid) {
      return await androidInfo.upgradeFromStore();
    } else if (Platform.isIOS) {
      return await iosInfo.upgradeFromStore();
    }
    return null;
  }

  Future<dynamic> showNormalDialog(BuildContext context) async {
    return await showDialog(
        context: context,
        barrierDismissible: isForce == null || isForce == false,
        builder: (BuildContext context) => RUpgradeNormalDialog(info: this));
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:r_upgrade_ui/r_upgrade_info.dart';
import 'package:r_upgrade_ui/r_upgrade_ui.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'RUpgrade UI Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with TickerProviderStateMixin {
  RUpgradeInfo rUpgradeInfo = RUpgradeInfo()
    ..androidInfo = (RUpgradeAndroidInfo()
      ..webUrl = 'http://www.baidu.com'
      ..downloadUrl =
          'https://mydata-1252536312.cos.ap-guangzhou.myqcloud.com/r_upgrade.apk'
      ..downloadFileName = 'r_upgrade.apk'
      ..downloadFinishAutoInstall = true
      ..downloadUseCache = false
      ..upgradeFlavor = RUpgradeFlavor.normal
      ..notificationVisibility = NotificationVisibility.VISIBILITY_VISIBLE
      ..notificationStyle = NotificationStyle.planTime)
    ..iosInfo = (RUpgradeIOSInfo()
      ..webUrl = 'http://www.google.com'
      ..isChina = false
      ..appId = '414478124')
    ..isForce = true
    ..newVersion = 'v1.0.1'
    ..newVersionCode = 1
    ..content = '1.fixed bug,\n2.fixed bug again'
    ..title = 'New Version';

  late AnimationController controller = AnimationController(
      vsync: this, duration: const Duration(milliseconds: 15));

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: ListView(
        children: [
          ListTile(
            onTap: () {
              rUpgradeInfo.showNormalDialog(context);
            },
            title: const Text('Normal Style'),
            trailing: const Icon(Icons.arrow_right_outlined),
          ),
          Padding(
            padding: const EdgeInsets.symmetric(horizontal: 8.0),
            child: AnimatedBuilder(
              animation: controller,
              builder: (BuildContext context, Widget? child) {
                return CustomPaint(
                  painter: AbsorbCustomPainter(
                      0.2, controller.value, 2, Colors.orange),
                  size: const Size(0, 30),
                );
              },
            ),
          ),
        ],
      ),
    );
  }

  @override
  void initState() {
    super.initState();
    controller.forward(from: 0.0);
  }
}

class AbsorbCustomPainter extends CustomPainter {
  final double speed;
  final double progress;
  final double total;
  final Color color;

  AbsorbCustomPainter(this.speed, this.progress, this.total, this.color);

  @override
  void paint(Canvas canvas, Size size) {
    canvas.drawRRect(
        RRect.fromRectAndRadius(Rect.fromLTWH(0, 0, size.width, size.height),
            const Radius.circular(8)),
        Paint()
          ..color = color
          ..strokeWidth = 2
          ..style = PaintingStyle.stroke);
    canvas.drawRRect(
        RRect.fromRectAndRadius(
            Rect.fromLTWH(0, 0, size.width * (progress / total), size.height),
            const Radius.circular(8)),
        Paint()
          ..color = color
          ..style = PaintingStyle.fill);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }
}
